src/tiledb/cloud/geospatial/ingestion.py

```diff
@@ -0,0 +1,1182 @@
+                    max_files=max_files,
+                )
+            else:
+                sources = find(
+                    search_uri,
+                    config=config,
+                    excludes=ignore,
+                    includes=raster_match,
+                    max_files=max_files,
+                )
+
+        if len(sources) == 0:
+            raise ValueError("No raster datasets found")
+
+        meta = get_raster_metadata(
+            sources, config=config, dst_tile_size=tile_size, res=res
+        )
+        kwargs.update(
+            {
+                "crs": meta.crs,
+                "extents": meta.extents,
+                "res": meta.res,
+                "band_count": meta.band_count,
+                "dtype": meta.dtype,
+                "nodata": nodata,
+                "resampling": resampling,
+                "tile_size": tile_size,
+                "compressor": compression_filter,
+            }
+        )
+        samples = meta.block_metadata
+    elif dataset_type == "geometry":
+        fn = ingest_geometry_udf
+
+        if search_uri:
+            if pattern:
+                sources = find(
+                    search_uri,
+                    config=config,
+                    excludes=ignore,
+                    includes=pattern,
+                    max_files=max_files,
+                )
+            else:
+                sources = find(
+                    search_uri,
+                    config=config,
+                    exclude=ignore,
+                    includes=geometry_match,
+                    max_files=max_files,
+                )
+
+        full_extents = get_geometry_metadata(sources, config=config)
+        kwargs.update(
+            {
+                "extents": full_extents,
+            }
+        )
+    else:
+        raise ValueError(f"Unsupported ingestion dataset type: {dataset_type}")
+
+    if not unit_testing:
+        # Build the task graph
+        dag_name = f"{dataset_type}-{DEFAULT_DAG_NAME}"
+        task_prefix = f"{dataset_type} - Task"
+
+        logger.info("Building graph")
+        graph = tiledb.cloud.dag.DAG(
+            name=dag_name,
+            mode=tiledb.cloud.dag.Mode.BATCH,
+            max_workers=workers,
+            namespace=namespace,
+        )
+
+        nodes = []
+
+        # create first node
+        ingest_node = graph.submit(
+            fn,
+            dataset_uri=dataset_uri,
+            config=config,
+            append=False,
+            verbose=verbose,
+            trace=trace,
+            log_uri=log_uri,
+            name=f"{task_prefix} - {0}",
+            mode=tiledb.cloud.dag.Mode.BATCH,
+            resources=DEFAULT_RESOURCES,
+            image_name=DEFAULT_IMG_NAME,
+            access_credentials_name=acn,
+            **kwargs,
+        )
+        # work functions
+        for i, work in enumerate(batch(samples, batch_size)):
+            logger.info(f"Adding batch {i + 1}")
```

Comment:
I would really advise the use of the `expand_node_output` approach instead of a for loop here.

Reply:
Why? `expand_node_output` takes the output from a previous node, the graph doesn't have that dependency.

Reply:
Yeap exactly. I had in mind that you can have a previous node in your taskgraph that splits the samples in batches and feed the outputs directly to the next node which is for ingestion. This way you avoid batching client side and `for loops` you leave the backend do the job for you and run all these tasks in parallel.

Reply:
👍